src/datadog/extraction_util.cpp

```diff
@@ -250,51 +251,49 @@ void AuditedReader::visit(
   if (found == contexts.end()) {
-    // Nothing extracted a trace ID. Return the first context that includes a
-    // parent ID, if any, or otherwise just return an empty `ExtractedData`.
-    // The purpose of looking for a parent ID is to allow for the error
-    // "extracted a parent ID without a trace ID," if that's what happened.
-    const auto other = std::find_if(
-        contexts.begin(), contexts.end(),
-        [](const ExtractedData& data) { return data.parent_id.has_value(); });
-    if (other != contexts.end()) {
-      result = *other;
-    }
     return result;
   }
 
   // `found` refers to the first extracted context that yielded a trace ID.
   // This will be our main context.
   //
-  // If the style of `found` is not W3C, then examine the remaining contexts
-  // for W3C-style tracestate that we might want to include in `result`.
-  result = *found;
-  if (result.style == PropagationStyle::W3C) {
-    return result;
-  }
+  // If the W3C style is present and its trace-id matches, we'll update the main
+  // context with tracestate information that we want to include in `result`. We
+  // may also need to use Datadog header information (only when the trace-id
+  // matches).
+  result = found->second;
 
-  const auto other =
-      std::find_if(found + 1, contexts.end(), [&](const ExtractedData& data) {
-        return data.style == PropagationStyle::W3C &&
-               data.trace_id == found->trace_id;
-      });
+  const auto w3c = contexts.find(PropagationStyle::W3C);
+  const auto dd = contexts.find(PropagationStyle::DATADOG);
 
-  if (other != contexts.end()) {
-    result.datadog_w3c_parent_id = other->datadog_w3c_parent_id;
-    result.additional_w3c_tracestate = other->additional_w3c_tracestate;
+  if (w3c != contexts.end() && w3c->second.trace_id == result.trace_id) {
+    result.additional_w3c_tracestate = w3c->second.additional_w3c_tracestate;
     result.additional_datadog_w3c_tracestate =
-        other->additional_datadog_w3c_tracestate;
+        w3c->second.additional_datadog_w3c_tracestate;
     result.headers_examined.insert(result.headers_examined.end(),
-                                   other->headers_examined.begin(),
-                                   other->headers_examined.end());
+                                   w3c->second.headers_examined.begin(),
+                                   w3c->second.headers_examined.end());
+
+    if (result.parent_id != w3c->second.parent_id) {
+      if (w3c->second.datadog_w3c_parent_id &&
+          w3c->second.datadog_w3c_parent_id != "0000000000000000") {
+        result.datadog_w3c_parent_id = w3c->second.datadog_w3c_parent_id;
+      } else if (dd != contexts.end() &&
+                 dd->second.trace_id == result.trace_id &&
+                 dd->second.parent_id.has_value()) {
+        result.datadog_w3c_parent_id = hex_padded(dd->second.parent_id.value());
+        // TODO: Do we add the result.headers_examined for all of the dd ones?
```

Comment:
@dmehala what should I do here? I left this TODO to be finished and removed before PR merge but wanted your input here

Reply:
`headers_examined` is mainly used for diagnostic purposes. Since this method is called `merge` I would argue the resulting `headers_examined` is the merge of all `headers_examined`.

The more I think about it, the more I think `headers_examined` should not be part `ExtractedData`. My advice would to not bother about it, a refactoring is needed.

Reply:
Sounds good, thanks